pkg/internal/cnsoperator/cnsfilevolumeclient/cnsfilevolumeclient.go

```diff
@@ -0,0 +1,200 @@
+// instance.
+// The instance is created if it doesn't exist.
+// Returns an error if the operation cannot be persisted on the API server.
+func (f *FileVolumeClients) AddClientVMToIPList(ctx context.Context, namespace, name, vmName, ip string) error {
+	log := logger.GetLogger(ctx)
+	log.Infof("Adding VM %s to CnsFileVolumeClient %s/%s list for IP address %s", vmName, namespace, name, ip)
+	actual, _ := f.fileVolumeClientsLock.LoadOrStore(namespace+"/"+name, &sync.Mutex{})
+	instanceLock, ok := actual.(*sync.Mutex)
+	if !ok {
+		return fmt.Errorf("failed to cast lock for cnsfilevolumeclient instance: %s/%s", namespace, name)
+	}
+	instanceLock.Lock()
+	defer instanceLock.Unlock()
+	obj, exists, err := f.store.GetByKey(namespace + "/" + name)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		// Create the instance as it does not exist on the API server.
+		instance := &v1alpha1.CnsFileVolumeClient{
+			ObjectMeta: v1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+			Spec: v1alpha1.CnsFileVolumeClientSpec{
+				ExternalIPtoClientVms: map[string][]string{
+					ip: {
+						vmName,
+					},
+				},
+			},
+		}
+		log.Debugf("Creating CnsFileVolumeClient instance %s/%s with spec: %+v", namespace, name, instance)
+		err = f.client.Create(ctx, instance)
+		if err != nil {
+			log.Errorf("failed to create CnsFileVolumeClient instance %s/%s with error: %+v", namespace, name, err)
+		}
+		return err
+	}
+	instance, ok := obj.(*v1alpha1.CnsFileVolumeClient)
+	if !ok {
+		return fmt.Errorf("unrecognized object: %v. Cannot cast to CnsFileVolumeClient type", obj)
+	}
+
+	// Verify if input VM exists in ExternalIPtoClientVms list for input IP address
+	log.Debugf("Verifying if VM %s exists in ExternalIPtoClientVms list for IP address: %s. Current list: %+v", vmName, ip, instance.Spec.ExternalIPtoClientVms[ip])
+	clientVMList := instance.Spec.ExternalIPtoClientVms[ip]
+	for _, clientVM := range clientVMList {
+		if clientVM == vmName {
+			log.Debugf("Found VM %s in list. Returning.", vmName)
+			return nil
+		}
+	}
+	clientVMList = append(clientVMList, vmName)
+	instance.Spec.ExternalIPtoClientVms[ip] = clientVMList
+	log.Debugf("Updating CnsFileVolumeClient instance %s/%s with spec: %+v", namespace, name, instance)
+	err = f.client.Update(ctx, instance)
+	if err != nil {
+		log.Errorf("failed to update CnsFileVolumeClient instance %s/%s with error: %+v", namespace, name, err)
+	}
+	return err
+}
+
+// RemoveClientVMFromIPList removes the input VM name, in the given namespace, from
+// the list of VM's that expose the same external IP address to the given CnsFileVolumeClient
+// instance.
+// If the given VM was the last client for this file volume, the instance is deleted from
+// the API server.
+// Returns an error if the operation cannot be persisted on the API server.
+func (f *FileVolumeClients) RemoveClientVMFromIPList(ctx context.Context, namespace, name, ip, vmName string) error {
```

Comment:
How about this method signature:

`RemoveFileVolumeClient(ctx context.Context, fileVolumeName, clientVMName, clientVMIP string) error`

Reply:
Done